File: Grind75_1/week6/SpiralMatrix.py
"""
https://leetcode.com/problems/spiral-matrix/
"""

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        rows = len(matrix)
        cols = len(matrix[0])

        visited = [[False] * cols for _ in range(rows)]

        dirs = [[0,1], [1,0], [0,-1], [-1,0]]

        resList = []

        r = 0
        c = -1
        dirIdx = 0

        for i in range (0, (rows * cols)):
            cnt = len(dirs)
            while (cnt >= 0):
                cnt = cnt - 1

                newR = r + dirs[dirIdx][0]
                newC = c + dirs[dirIdx][1]
                
                if (newR < 0 or newR >= rows or newC < 0 or newC >= cols or visited[newR][newC] == True):
                    if (dirIdx + 1 < len(dirs)):
                        dirIdx = dirIdx + 1
                    else:
                        dirIdx = 0
                    continue
                elif (visited[newR][newC] == False):
                    r = newR
                    c = newC
                    break
                    

            resList.append(matrix[r][c])
            visited[r][c] = True

        return resList

File: neetcode/blind75_JS/Math-and-Geometry/spiral-matrix.js
// https://neetcode.io/problems/spiral-matrix

class Solution {
    /**
     * @param {number[][]} matrix
     * @return {number[]}
     */
    spiralOrder(matrix) {
        if (matrix === null) {
            return []
        }

        const res = []

        let l = 0
        let r = matrix[0].length
        let t = 0
        let b = matrix.length

        while (l < r && t < b) {
            for (let i = l; i < r; i ++) {
                res.push(matrix[t][i])
            }
            t += 1

            for (let i = t; i < b; i ++) {
                res.push(matrix[i][r - 1])
            }
            r -= 1

            if (!(l < r && t < b)) {
                break;
            }

            for (let i = r-1; i >= l; i --) {
                res.push(matrix[b - 1][i])
            }
            b -= 1

            for (let i = b-1; i >= t; i --) {
                res.push(matrix[i][l])
            }
            l += 1
        }

        return res
    }
}
